File: connectors/go/apiglimpse/middleware.go
package apiglimpse

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"math/rand"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Config controls the chi/net/http middleware.
type Config struct {
	AgentURL  string
	APIKey    string
	// ServiceName labels this app as a topology caller (API_SENSOR_SERVICE_NAME).
	ServiceName string
	// SampleRate is 0–1. nil means default 1.0 (sample everything).
	SampleRate *float64
	FlushInterval           time.Duration
	MaxBatchSize            int
	MaxBufferSize           int
	RequestTimeout          time.Duration
	CircuitFailureThreshold int
	CircuitOpenFor          time.Duration
	// HTTPClient optional override (tests).
	HTTPClient *http.Client
}

// ConfigFromEnv builds Config from API_SENSOR_* environment variables.
func ConfigFromEnv() Config {
	cfg := DefaultConfig()
	if v := os.Getenv("API_SENSOR_AGENT_URL"); v != "" {
		cfg.AgentURL = v
	}
	if v := os.Getenv("API_SENSOR_KEY"); v != "" {
		cfg.APIKey = v
	}
	if v := os.Getenv("API_SENSOR_SERVICE_NAME"); v != "" {
		cfg.ServiceName = v
	}
	if v := os.Getenv("API_SENSOR_SAMPLE_RATE"); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			cfg.SampleRate = &f
		}
	}
	return cfg
}

// DefaultConfig returns Express-middleware-aligned defaults.
func DefaultConfig() Config {
	rate := 1.0
	return Config{
		AgentURL:                "http://localhost:8080",
		APIKey:                  "",
		SampleRate:              &rate,
		FlushInterval:           time.Second,
		MaxBatchSize:            50,
		MaxBufferSize:           500,
		RequestTimeout:          2 * time.Second,
		CircuitFailureThreshold: 3,
		CircuitOpenFor:          15 * time.Second,
	}
}

type sensor struct {
	cfg       Config
	mu        sync.Mutex
	buffer    []Sample
	flushing  bool
	failures  int
	openUntil time.Time
	client    *http.Client
	stopCh    chan struct{}
	rate      float64
}

// Middleware returns chi-compatible middleware that samples traffic and
// asynchronously POSTs envelope v1 to the hosted agent. Fail-open: never
// blocks or fails the customer request because of API Glimpse.
func Middleware(cfg Config) func(http.Handler) http.Handler {
	c := mergeConfig(DefaultConfig(), cfg)

	rate := 1.0
	if c.SampleRate != nil {
		rate = *c.SampleRate
	}

	s := &sensor{
		cfg:    c,
		rate:   rate,
		buffer: make([]Sample, 0, c.MaxBatchSize),
		stopCh: make(chan struct{}),
		client: c.HTTPClient,
	}
	if s.client == nil {
		s.client = &http.Client{Timeout: c.RequestTimeout}
	}

	go s.loop()

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !s.shouldSample() {
				next.ServeHTTP(w, r)
				return
			}

			start := time.Now()
			var reqBody []byte
			func() {
				defer func() { _ = recover() }()
				reqBody, r = peekRequestBody(r)
			}()

			rw := &captureWriter{ResponseWriter: w, status: http.StatusOK}
			next.ServeHTTP(rw, r)

			func() {
				defer func() { _ = recover() }()
				s.enqueue(r, rw, reqBody, start)
			}()
		})
	}
}

func mergeConfig(base, over Config) Config {
	if over.AgentURL != "" {
		base.AgentURL = over.AgentURL
	}
	if over.APIKey != "" {
		base.APIKey = over.APIKey
	}
	if over.ServiceName != "" {
		base.ServiceName = over.ServiceName
	}
	if over.SampleRate != nil {
		base.SampleRate = over.SampleRate
	}
	if over.FlushInterval > 0 {
		base.FlushInterval = over.FlushInterval
	}
	if over.MaxBatchSize > 0 {
		base.MaxBatchSize = over.MaxBatchSize
	}
	if over.MaxBufferSize > 0 {
		base.MaxBufferSize = over.MaxBufferSize
	}
	if over.RequestTimeout > 0 {
		base.RequestTimeout = over.RequestTimeout
	}
	if over.CircuitFailureThreshold > 0 {
		base.CircuitFailureThreshold = over.CircuitFailureThreshold
	}
	if over.CircuitOpenFor > 0 {
		base.CircuitOpenFor = over.CircuitOpenFor
	}
	if over.HTTPClient != nil {
		base.HTTPClient = over.HTTPClient
	}
	return base
}

func (s *sensor) loop() {
	t := time.NewTicker(s.cfg.FlushInterval)
	defer t.Stop()
	for {
		select {
		case <-t.C:
			s.flush()
		case <-s.stopCh:
			return
		}
	}
}

func (s *sensor) shouldSample() bool {
	if s.rate >= 1 {
		return true
	}
	if s.rate <= 0 {
		return false
	}
	return rand.Float64() < s.rate
}

func (s *sensor) enqueue(r *http.Request, rw *captureWriter, reqBody []byte, start time.Time) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.buffer) >= s.cfg.MaxBufferSize {
		s.buffer = s.buffer[1:]
	}
	reqHeaders := flattenHeader(r.Header)
	resHeaders := flattenHeader(rw.Header())
	var reqParsed any
	hasReq := false
	if len(reqBody) > 0 {
		if err := json.Unmarshal(reqBody, &reqParsed); err == nil {
			hasReq = true
		}
	}
	var resParsed any
	hasRes := false
	resCT := strings.ToLower(resHeaders["content-type"])
	// Prefer JSON Content-Type; still accept bodies that parse as JSON when CT unset.
	// Skip obvious binary / streaming content types.
	skipRes := strings.Contains(resCT, "octet-stream") ||
		strings.Contains(resCT, "event-stream") ||
		strings.HasPrefix(resCT, "image/") ||
		strings.HasPrefix(resCT, "audio/") ||
		strings.HasPrefix(resCT, "video/") ||
		strings.Contains(resCT, "multipart/")
	if !skipRes && len(rw.body) > 0 {
		if resCT == "" || strings.Contains(resCT, "application/json") || strings.Contains(resCT, "+json") {
			if err := json.Unmarshal(rw.body, &resParsed); err == nil {
				hasRes = true
			}
		}
	}
	path := r.URL.Path
	if path == "" {
		path = "/"
	}
	captured := hasRes
	sample := CreateSample(SampleInput{
		Method:               r.Method,
		Path:                 path,
		StatusCode:           rw.status,
		LatencyMs:            time.Since(start).Milliseconds(),
		RequestHeaders:       reqHeaders,
		ResponseHeaders:      resHeaders,
		RequestHeaderNames:   headerNamesFromHTTP(r.Header),
		ResponseHeaderNames:  headerNamesFromHTTP(rw.Header()),
		RequestBody:          reqParsed,
		ResponseBody:         resParsed,
		HasRequestBody:       hasReq,
		HasResponseBody:      hasRes,
		ResponseBodyCaptured: &captured,
		Caller:               ResolveCaller(reqHeaders, s.cfg.ServiceName),
		AuthObserved:         ObserveAuth(reqHeaders),
	})
	s.buffer = append(s.buffer, sample)
	if len(s.buffer) >= s.cfg.MaxBatchSize {
		go s.flush()
	}
}

func (s *sensor) recordFailure() {
	s.failures++
	if s.failures >= s.cfg.CircuitFailureThreshold {
		s.openUntil = time.Now().Add(s.cfg.CircuitOpenFor)
		s.failures = 0
	}
}

func (s *sensor) recordSuccess() {
	s.failures = 0
	s.openUntil = time.Time{}
}

func (s *sensor) flush() {
	s.mu.Lock()
	if s.flushing || len(s.buffer) == 0 || time.Now().Before(s.openUntil) {
		s.mu.Unlock()
		return
	}
	s.flushing = true
	n := s.cfg.MaxBatchSize
	if n > len(s.buffer) {
		n = len(s.buffer)
	}
	batch := make([]Sample, n)
	copy(batch, s.buffer[:n])
	s.buffer = s.buffer[n:]
	apiKey := s.cfg.APIKey
	agentURL := strings.TrimRight(s.cfg.AgentURL, "/")
	timeout := s.cfg.RequestTimeout
	client := s.client
	s.mu.Unlock()

	defer func() {
		s.mu.Lock()
		s.flushing = false
		s.mu.Unlock()
	}()

	env := CreateEnvelope(apiKey, batch)
	body, err := json.Marshal(env)
	if err != nil {
		s.mu.Lock()
		s.recordFailure()
		s.mu.Unlock()
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, agentURL+"/v1/samples", bytes.NewReader(body))
	if err != nil {
		s.mu.Lock()
		s.recordFailure()
		s.mu.Unlock()
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-API-Key", apiKey)

	res, err := client.Do(req)
	if err != nil {
		s.mu.Lock()
		s.recordFailure()
		s.mu.Unlock()
		return
	}
	defer res.Body.Close()
	_, _ = io.Copy(io.Discard, res.Body)

	s.mu.Lock()
	defer s.mu.Unlock()
	if res.StatusCode >= 500 {
		s.recordFailure()
		return
	}
	if res.StatusCode == http.StatusUnauthorized {
		return
	}
	s.recordSuccess()
}

type captureWriter struct {
	http.ResponseWriter
	status int
	body   []byte
}

func (w *captureWriter) WriteHeader(code int) {
	w.status = code
	w.ResponseWriter.WriteHeader(code)
}

func (w *captureWriter) Write(b []byte) (int, error) {
	if len(w.body) < 64*1024 {
		remain := 64*1024 - len(w.body)
		if len(b) <= remain {
			w.body = append(w.body, b...)
		} else {
			w.body = append(w.body, b[:remain]...)
		}
	}
	return w.ResponseWriter.Write(b)
}

func peekRequestBody(r *http.Request) ([]byte, *http.Request) {
	if r.Body == nil {
		return nil, r
	}
	const max = 64 * 1024
	buf, err := io.ReadAll(io.LimitReader(r.Body, max+1))
	_ = r.Body.Close()
	if err != nil {
		r.Body = http.NoBody
		return nil, r
	}
	if len(buf) > max {
		r.Body = io.NopCloser(bytes.NewReader(buf))
		return nil, r
	}
	r.Body = io.NopCloser(bytes.NewReader(buf))
	return buf, r
}

func flattenHeader(h http.Header) map[string]string {
	out := make(map[string]string, len(h))
	for k, vals := range h {
		out[strings.ToLower(k)] = strings.Join(vals, ", ")
	}
	return out
}

func headerNamesFromHTTP(h http.Header) []string {
	names := make([]string, 0, len(h))
	for k := range h {
		names = append(names, strings.ToLower(k))
	}
	return names
}

File: connectors/python/src/apiglimpse/middleware.py
"""FastAPI / Starlette middleware — fail-open async flush + circuit breaker."""

from __future__ import annotations

import asyncio
import json
import os
import random
import time
from typing import Any, Callable

import httpx
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from starlette.types import ASGIApp

from .envelope import create_envelope, create_sample

DEFAULTS = {
    "agent_url": "http://localhost:8080",
    "api_key": "",
    "sample_rate": 1.0,
    "flush_interval_ms": 1000,
    "max_batch_size": 50,
    "max_buffer_size": 500,
    "request_timeout_ms": 2000,
    "circuit_failure_threshold": 3,
    "circuit_open_ms": 15000,
}


def _env_float(name: str, default: float) -> float:
    raw = os.environ.get(name)
    if raw is None or raw == "":
        return default
    try:
        return float(raw)
    except ValueError:
        return default


def _observe_auth(headers: dict[str, str]) -> str:
    try:
        auth = headers.get("authorization") or headers.get("Authorization")
        if auth and str(auth).lower().startswith("bearer "):
            return "bearer"
        if headers.get("cookie") or headers.get("Cookie"):
            return "cookie"
        return "none"
    except Exception:
        return "none"


def _headers_dict(headers: Any) -> dict[str, str]:
    try:
        return {str(k): str(v) for k, v in headers.items()}
    except Exception:
        return {}


def _parse_json_bytes(raw: bytes | None) -> Any:
    if not raw:
        return None
    try:
        return json.loads(raw.decode("utf-8"))
    except Exception:
        return None


class ApiGlimpseMiddleware(BaseHTTPMiddleware):
    """
    Capture request/response metadata, redact secrets, buffer samples, and
    POST envelope v1 to ``{agent_url}/v1/samples`` with ``X-API-Key``.

    Never blocks or fails the customer request because of API Glimpse.
    """

    def __init__(
        self,
        app: ASGIApp,
        *,
        agent_url: str | None = None,
        api_key: str | None = None,
        sample_rate: float | None = None,
        flush_interval_ms: int | None = None,
        max_batch_size: int | None = None,
        max_buffer_size: int | None = None,
        request_timeout_ms: int | None = None,
        circuit_failure_threshold: int | None = None,
        circuit_open_ms: int | None = None,
    ) -> None:
        super().__init__(app)

        env_agent = os.environ.get("API_SENSOR_AGENT_URL") or DEFAULTS["agent_url"]
        env_key = os.environ.get("API_SENSOR_KEY") or DEFAULTS["api_key"]
        env_rate = _env_float("API_SENSOR_SAMPLE_RATE", DEFAULTS["sample_rate"])

        self.agent_url = (agent_url if agent_url is not None else env_agent).rstrip("/")
        self.api_key = api_key if api_key is not None else env_key
        self.sample_rate = sample_rate if sample_rate is not None else env_rate
        self.flush_interval_ms = (
            flush_interval_ms if flush_interval_ms is not None else DEFAULTS["flush_interval_ms"]
        )
        self.max_batch_size = (
            max_batch_size if max_batch_size is not None else DEFAULTS["max_batch_size"]
        )
        self.max_buffer_size = (
            max_buffer_size if max_buffer_size is not None else DEFAULTS["max_buffer_size"]
        )
        self.request_timeout_ms = (
            request_timeout_ms
            if request_timeout_ms is not None
            else DEFAULTS["request_timeout_ms"]
        )
        self.circuit_failure_threshold = (
            circuit_failure_threshold
            if circuit_failure_threshold is not None
            else DEFAULTS["circuit_failure_threshold"]
        )
        self.circuit_open_ms = (
            circuit_open_ms if circuit_open_ms is not None else DEFAULTS["circuit_open_ms"]
        )

        self._buffer: list[dict[str, Any]] = []
        self._flushing = False
        self._consecutive_failures = 0
        self._circuit_open_until = 0.0
        self._flush_task: asyncio.Task[None] | None = None
        self._lock = asyncio.Lock()
        self._started = False

    def _should_sample(self) -> bool:
        if self.sample_rate >= 1:
            return True
        if self.sample_rate <= 0:
            return False
        return random.random() < self.sample_rate

    def _circuit_open(self) -> bool:
        return time.time() * 1000 < self._circuit_open_until

    def _record_failure(self) -> None:
        self._consecutive_failures += 1
        if self._consecutive_failures >= self.circuit_failure_threshold:
            self._circuit_open_until = time.time() * 1000 + self.circuit_open_ms
            self._consecutive_failures = 0

    def _record_success(self) -> None:
        self._consecutive_failures = 0
        self._circuit_open_until = 0.0

    def _ensure_flusher(self) -> None:
        if self._started:
            return
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            return
        self._started = True
        self._flush_task = loop.create_task(self._flush_loop())

    async def _flush_loop(self) -> None:
        interval = max(self.flush_interval_ms, 50) / 1000.0
        while True:
            try:
                await asyncio.sleep(interval)
                await self._flush()
            except asyncio.CancelledError:
                break
            except Exception:
                # fail-open: never crash the flusher permanently
                continue

    async def _flush(self) -> None:
        if self._flushing or not self._buffer:
            return
        if self._circuit_open():
            return

        self._flushing = True
        async with self._lock:
            batch = self._buffer[: self.max_batch_size]
            del self._buffer[: self.max_batch_size]

        try:
            envelope = create_envelope(api_key=self.api_key, samples=batch)
            timeout = self.request_timeout_ms / 1000.0
            url = f"{self.agent_url}/v1/samples"
            async with httpx.AsyncClient(timeout=timeout) as client:
                res = await client.post(
                    url,
                    json=envelope,
                    headers={
                        "Content-Type": "application/json",
                        "X-API-Key": self.api_key or "",
                    },
                )
            if not res.is_success and res.status_code >= 500:
                self._record_failure()
            elif not res.is_success and res.status_code == 401:
                # bad key — drop, do not trip circuit forever
                pass
            else:
                self._record_success()
        except Exception:
            self._record_failure()
        finally:
            self._flushing = False

    def _enqueue(self, sample: dict[str, Any]) -> None:
        if len(self._buffer) >= self.max_buffer_size:
            self._buffer.pop(0)
        self._buffer.append(sample)
        if len(self._buffer) >= self.max_batch_size:
            try:
                loop = asyncio.get_running_loop()
                loop.create_task(self._flush())
            except RuntimeError:
                pass

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        self._ensure_flusher()

        try:
            if not self._should_sample():
                return await call_next(request)
        except Exception:
            return await call_next(request)

        start = time.time()
        request_body: Any = None
        response_body: Any = None

        try:
            content_type = (request.headers.get("content-type") or "").lower()
            if "application/json" in content_type:
                raw = await request.body()
                request_body = _parse_json_bytes(raw)

                async def receive() -> dict[str, Any]:
                    return {"type": "http.request", "body": raw, "more_body": False}

                request = Request(request.scope, receive)
        except Exception:
            request_body = None

        try:
            response = await call_next(request)
        except Exception:
            # Never mask app errors; still no sample if we couldn't get a response
            raise

        try:
            # Capture JSON response body without breaking streaming consumers
            resp_ct = (response.headers.get("content-type") or "").lower()
            if "application/json" in resp_ct and hasattr(response, "body_iterator"):
                chunks: list[bytes] = []
                async for chunk in response.body_iterator:
                    if isinstance(chunk, str):
                        chunks.append(chunk.encode("utf-8"))
                    else:
                        chunks.append(chunk)
                body_bytes = b"".join(chunks)
                response_body = _parse_json_bytes(body_bytes)

                async def body_iter():
                    yield body_bytes

                response = Response(
                    content=body_bytes,
                    status_code=response.status_code,
                    headers=dict(response.headers),
                    media_type=response.media_type,
                    background=response.background,
                )

            latency_ms = (time.time() - start) * 1000.0
            path = request.url.path or "/"
            sample = create_sample(
                method=request.method,
                path=path,
                status_code=response.status_code,
                latency_ms=latency_ms,
                request_headers=_headers_dict(request.headers),
                response_headers=_headers_dict(response.headers),
                request_body=request_body,
                response_body=response_body,
                auth_observed=_observe_auth(_headers_dict(request.headers)),
            )
            self._enqueue(sample)
        except Exception:
            # fail-open: never break the app
            pass

        return response
